### PR TITLE
Implement a Backward Compatibility New Error Model for training Tensorflow models using the BC New Error Loss

### PR DESCRIPTION
**Pull request by ilmarinen**
The implementation in Tensorflow is different from the Pytorch implementation as follows:

1. In order to allow the models to be efficiently trained (without eager execution) using the `.fit()` method, we implemented a custom training loop on the model via the `.train_step()` method.

2. If the user has an existing model `h1` and would like to train a new model `h2`, the new model must be instantiated as follows:

```
lambda_c = 0.5
h2 = BCNewErrorCompatibilityModel([ ... keras layers ...], h1=h1, lambda_c=lambda_c)
```

3. And then one may train it using:

```
h2.compile(
    loss=tf.keras.losses.sparse_categorical_crossentropy,
    optimizer=tf.keras.optimizers.Adam(0.001),
    metrics=['accuracy']
)

h2.fit(
    ds_train,
    epochs=6,
    validation_data=ds_test,
)
```

The model `h2` is trained using the Backward Compatibility loss function using the loss function specified by the user in the `.compile()` method above.

### CODE REVIEW

File: backwardcompatibilityml/tensorflow/models.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return cross_entropy_loss
+
+    def loss_func(self, x, y, loss=None):
+        """
+        Backward compatibility loss function to be used by the model
+        """
+        if loss is None:
+            loss = tf.keras.losses.SparseCategoricalCrossentropy(
+                reduction=tf.keras.losses.Reduction.SUM)
+        h1_output = tf.argmax(self.h1(x), axis=1)
```
Comment (Nking92) [inline backwardcompatibilityml/tensorflow/models.py:38]: This fails if h1=None. Are you sure you want to allow h1=None in the constructor?

**Review by Nking92 — APPROVED** [review]
1 small potential issue, otherwise looks good. I was able to run the notebook on my machine and it worked fine.

File: backwardcompatibilityml/tensorflow/models.py (hunk 1 above)
Comment (ilmarinen) [reply backwardcompatibilityml/tensorflow/models.py:38]: Good point. Though I can't think of any scenario where the user would not set `h1`.
What I can do is just raise an exception at instantiation time if the user sets it to None. And make sure to document that we need that.

**Review by nushib — APPROVED** [review]

File: backwardcompatibilityml/tensorflow/helpers.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            # on the GradientTape.
+            # Compute the loss value for this minibatch.
+            loss_value = bc_loss(x_batch_train, y_batch_train)
+
+        # Use the gradient tape to automatically retrieve
+        # the gradients of the trainable variables with respect to the loss.
+        grads = tape.gradient(loss_value, h2.trainable_weights)
+
+        # Run one step of gradient descent by updating
+        # the value of the variables to minimize the loss.
```
Comment (Nking92) [inline backwardcompatibilityml/tensorflow/helpers.py:23]: Thanks for leaving well commented code

File: examples/tensorflow-MNIST-generalized.ipynb [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class BCCrossEntropyLoss(object):\n",
```
Comment (Nking92) [inline examples/tensorflow-MNIST-generalized.ipynb:119]: Is there a reason why you chose to copy-paste this class instead of importing it?

Comment (ilmarinen) [reply examples/tensorflow-MNIST-generalized.ipynb:119]: Good catch. This was leftover.

Comment (ilmarinen) [reply examples/tensorflow-MNIST-generalized.ipynb:119]: Fixed and pushed changes.

**Review by jlema — APPROVED** [review]
